go/oasis-node/cmd/debug/txsource/workload/delegation.go

```diff
@@ -0,0 +1,263 @@
+package workload
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"google.golang.org/grpc"
+
+	"github.com/oasislabs/oasis-core/go/common/crypto/signature"
+	memorySigner "github.com/oasislabs/oasis-core/go/common/crypto/signature/signers/memory"
+	"github.com/oasislabs/oasis-core/go/common/logging"
+	consensus "github.com/oasislabs/oasis-core/go/consensus/api"
+	"github.com/oasislabs/oasis-core/go/consensus/api/transaction"
+	runtimeClient "github.com/oasislabs/oasis-core/go/runtime/client/api"
+	staking "github.com/oasislabs/oasis-core/go/staking/api"
+)
+
+const (
+	// NameDelegation is the name of the delegation workload.
+	NameDelegation = "delegation"
+
+	delegationNumAccounts = 10
+	delegateAmount        = 100
+)
+
+var delegationLogger = logging.GetLogger("cmd/txsource/workload/delegation")
+
+type delegation struct{}
+
+func (delegation) Run(gracefulExit context.Context, rng *rand.Rand, conn *grpc.ClientConn, cnsc consensus.ClientBackend, rtc runtimeClient.RuntimeClient, fundingAccount signature.Signer) error { // nolint: gocyclo
+	var err error
+	ctx := context.Background()
+
+	fac := memorySigner.NewFactory()
+	accounts := make([]struct {
+		signer              signature.Signer
+		reckonedNonce       uint64
+		delegatedTo         signature.PublicKey
+		debondingUntilEpoch uint64
+	}, delegationNumAccounts)
+
+	for i := range accounts {
+		accounts[i].signer, err = fac.Generate(signature.SignerEntity, rng)
+		if err != nil {
+			return fmt.Errorf("memory signer factory Generate account %d: %w", i, err)
+		}
+
+		// Fund the account with delegation amount.
+		// Funds for fee's will be transferred before making transactions.
+		if err = transferFunds(ctx, delegationLogger, cnsc, fundingAccount, accounts[i].signer.Public(), delegateAmount); err != nil {
+			return fmt.Errorf("account funding failure: %w", err)
+		}
+	}
+
+	client := staking.NewStakingClient(conn)
+
+MAIN:
+	for {
+		// There are multiple loop branches so check for termination here.
+		select {
+		case <-time.After(1 * time.Second):
+		case <-gracefulExit.Done():
+			delegationLogger.Debug("time's up")
+			return nil
+		}
+
+		// Get current epoch.
+		epoch, err := cnsc.GetEpoch(ctx, consensus.HeightLatest)
+		if err != nil {
+			return fmt.Errorf("GetEpoch: %w", err)
+		}
+
+		var signedTx *transaction.SignedTransaction
+		var gas transaction.Gas
+		switch rng.Intn(2) {
```

Comment:
for maintainability, consider refactoring each case into a helper function

Reply:
done